orttraining/orttraining/core/optimizer/memory_optimizer/common.cc

```diff
@@ -0,0 +1,312 @@
+ *   - value: a pair of bool, representing whether the activation is used by forward nodes or by backward nodes.
+ * @param is_forward_nodes Collected node is forward pass op mapping.
+ */
+void GetForwardOutputUsageMap(const Graph& graph,
+                              const ptrdiff_t boundary_op_order_in_topological_sort,
+                              const InlinedHashMap<NodeIndex, size_t>&
+                                  node_index_to_its_order_in_topological_sort_map,
+                              InlinedHashMap<std::string, std::pair<bool, bool>>& fw_op_output_arg_used_map,
+                              InlinedHashMap<const Node*, bool>& is_forward_nodes) {
+  fw_op_output_arg_used_map.clear();
+  ORT_ENFORCE(boundary_op_order_in_topological_sort >= 0);
+  GraphViewer graph_viewer(graph);
+  const auto& node_ids = graph_viewer.GetNodesInTopologicalOrder();
+  is_forward_nodes.reserve(node_ids.size());
+
+  auto is_forward_pass_operator = [](ptrdiff_t op_order_in_topological_sort,
+                                     ptrdiff_t boundary_op_order_in_topological_sort) -> bool {
+    return op_order_in_topological_sort <= boundary_op_order_in_topological_sort;
+  };
+
+  for (size_t i = 0; i < node_ids.size(); ++i) {
+    const Node* p_node = graph.GetNode(node_ids[i]);
+    if (p_node == nullptr /* skip removed nodes*/) {
+      continue;
+    }
+
+    const Node& node = *p_node;
+
+    bool is_forward_op = is_forward_pass_operator(static_cast<ptrdiff_t>(i), boundary_op_order_in_topological_sort);
+    if (!is_forward_op) {
+      is_forward_nodes[p_node] = false;
+      continue;
+    }
+
+    is_forward_nodes[p_node] = true;
+
+    for (auto& output_arg : node.OutputDefs()) {
+      bool used_in_fw = false;
+      bool used_in_bw = false;
+      for (auto& consumer_node : graph.GetConsumerNodes(output_arg->Name())) {
+        size_t consumer_node_index_in_topological_order =
+            node_index_to_its_order_in_topological_sort_map.at(consumer_node->Index());
+        if (is_forward_pass_operator(static_cast<ptrdiff_t>(consumer_node_index_in_topological_order),
+                                     boundary_op_order_in_topological_sort)) {
+          used_in_fw = true;
+        } else {
+          used_in_bw = true;
+        }
+      }
+      fw_op_output_arg_used_map.insert({{output_arg->Name(), std::make_pair(used_in_fw, used_in_bw)}});
+    }
+  }
+}
+
+// trim from start (in place)
+static inline void ltrim(std::string& s) {
+  s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
+            return !std::isspace(ch);
+          }));
+}
+
+// trim from end (in place)
+static inline void rtrim(std::string& s) {
+  s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
+            return !std::isspace(ch);
+          }).base(),
+          s.end());
+}
+
+// trim from both ends
+static inline std::string trim_copy(std::string s) {
+  rtrim(s);
+  ltrim(s);
+  return s;
+}
```

Comment:
I can see some other code can benefit from this?
Suggest moving it to the header or even in the ORT common header. May be string utils;